android/greengrass-nucleus-android/src/main/java/com/aws/greengrass/android/service/DefaultGreengrassComponentService.java

```diff
@@ -0,0 +1,375 @@
+                || REQUEST_RESTART == exitCode || REQUEST_REBOOT == exitCode) {
+            errorDetected = false;
+        }
+    }
+
+    /**
+     * Resets start attempts counter back to zero.
+     */
+    public static void resetStartAttemptsCounter() {
+        synchronized (startAttemptsCounter) {
+            startAttemptsCounter = 0;
+        }
+    }
+
+    /**
+     * Starting Nucleus as Android Foreground Service.
+     *
+     * @param context context of application
+     */
+    public static synchronized void launch(Context context) {
+        initLogger(context);
+        if (componentManager == null) {
+            final AndroidContextProvider androidContextProvider = new AndroidContextProvider() {
+                // FIXME: get return reference to dead context ?
+                @Override
+                public Context getContext() {
+                    return context;
+                }
+            };
+
+            componentManager = new AndroidBaseComponentManager(androidContextProvider);
+        }
+
+        HashMap<String, String> environment = new HashMap<String, String>();
+        environment.put("SVCUID", authToken);
+
+        Thread startupThread = new Thread(() -> {
+            try {
+                componentManager.startService(context.getPackageName(),
+                        DefaultGreengrassComponentService.class.getCanonicalName(),
+                        ACTION_START_COMPONENT,
+                        null,
+                        environment,
+                        logger,
+                        s -> {
+                            String ss = s.toString().trim();
+                            logger.atInfo().setEventType("stdout").log(ss);
+                        },
+                        s -> {
+                            String ss = s.toString().trim();
+                            logger.atWarn().setEventType("stderr").log(ss);
+                        });
+            } catch (Throwable e) {
+                logger.atError().setCause(e).log("Couldn't start Nucleus service");
+            }
+        });
+        startupThread.setDaemon(true);
+        startupThread.start();
+    }
+
+    /**
+     * Stop Nucleus as Android Foreground Service.
+     *
+     * @param context context of application
+     */
+    public static void finish(Context context) {
+        if (componentManager != null) {
+            try {
+                componentManager.stopService(context.getPackageName(),
+                        DefaultGreengrassComponentService.class.getCanonicalName(),
+                        logger);
+            } catch (Throwable e) {
+                logger.atError().setCause(e).log("Couldn't stop Nucleus service");
+            }
+        }
+    }
+
+    /**
+     * Schedules restart of Nucleus.
+     *
+     * @param dueToError true when error occured
+     */
+    public void scheduleRestart(boolean dueToError) {
+        synchronized (startAttemptsCounter) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Object startAttemptsCounter synchronized on line 264 should not be re-assigned in the same synchronized block on line 267. After this re-assignment, you lose the reference to the synchronized object.